cluster-autoscaler/simulator/predicates.go

```diff
@@ -261,7 +228,16 @@ func (p *PredicateChecker) GetPredicateMetadata(pod *apiv1.Pod, nodeInfos map[st
 	if !p.enableAffinityPredicate {
 		return nil
 	}
-	return p.predicateMetadataProducer(pod, nodeInfos)
+	nodeInfoList := make([]*schedulernodeinfo.NodeInfo, 0, len(nodeInfos))
```

Comment:
how is `nodeInfos` created? Maybe you should try to use `NewSnapshot` or provide your own implementation of `listers.SharedLister`
This usage of snapshot seems to be relying too much on internals.

Reply:
Multiple versions of `nodeInfos` are created as part of scale-down simulation here: https://github.com/kubernetes/autoscaler/blob/dc64d0aab25202144cf655c8526d5d6c2d7e5f97/cluster-autoscaler/simulator/cluster.go#L221.

It does not seem feasible to migrate the code to `NewSnapshot`. If possible at all major refactor of the method would be needed. 

Do you think I should create our own implementation of `SharedLister`? It would be a copy of what you have. Yet I agree it will improve the separation of CA and scheduler.

Reply:
middle ground: if CA wants to maintain the ```NodeInfoMap```, then we can have a ```NewSnapshotFromNodeInfoMap``` func to create the snapshot. 

Note that we do a little more than just create the ```NodeInfoList```, but also a ```HavePodsWithAffinityNodeInfoList```:  https://github.com/kubernetes/kubernetes/blob/bcb171b375df0dd9326bf5d3a23428beea1aeb89/pkg/scheduler/nodeinfo/snapshot/snapshot.go#L52

Reply:
Once https://github.com/kubernetes/kubernetes/pull/85139 is merged, you should be able to use NewSnapshot.

Reply:
Thanks! Updated the PR.

Reply:
wait for this one as well https://github.com/kubernetes/kubernetes/pull/85151 , should allow you to not specify algorithm source explicitly